reconcile/vault_replication.py

```diff
@@ -0,0 +1,225 @@
+import logging
+import re
+
+from reconcile.utils.vault import VaultClient, _VaultClient, SecretNotFound
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+from reconcile.gql_definitions.jenkins_configs.jenkins_configs import (
+    JenkinsConfigV1,
+)
+
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultReplicationV1,
+    VaultInstanceV1,
+)
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPolicyV1,
+)
+
+from reconcile.utils import gql
+
+from typing import List, cast, Optional
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault,
+    dest_vault,
+    current_dest_version,
+    current_source_version,
+    path,
+):
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+        copy_secret(dry_run, source_vault, dest_vault, path, version, secret_dict)
+
+
+def copy_secret(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    path: str,
+    version: int,
+    secret_dict: dict,
+) -> None:
+    secret, version = source_vault.read_all_version(secret_dict)
+    write_dict = {"path": path, "data": secret}
+    logging.info(["replicate_vault_secret", version, path])
+    if not dry_run:
+        dest_vault.write(write_dict)
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: _VaultClient, dest_vault: _VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    _, version = source_vault.read_all_version(secret_dict)
+
+    try:
+        _, dest_version = dest_vault.read_all_version(secret_dict)
+        if dest_version is None and version is None:
+            # Secret is a v1 and does not return version
+            copy_secret(dry_run, source_vault, dest_vault, path, 1, secret_dict)
+        elif dest_version < version:
+            deep_copy_versions(
+                dry_run, source_vault, dest_vault, dest_version, version, path
+            )
+        else:
+            logging.info(["replicate_vault_secret", dest_version, version, path])
+    except SecretNotFound:
+        logging.info(["replicate_vault_secret", "Secret not found", path])
+        deep_copy_versions(dry_run, source_vault, dest_vault, 0, version, path)
+
+
+def check_copy_secret_list(
```

Comment:
nit: I don't really like functions that do two different things. I'd rather see one function to check and one function to copy. That would even make clearer that we won't move on if there are paths the approle cannot copy.

Reply:
Simplified the function changing it for two different with clear functionality